### PR TITLE
CI services chapter (GitHub actions)

### PR DESCRIPTION
**Pull request by BatoolMM**
### Summary

This is part of a new sub-chapter on GitHub action, @bsipocz and I are working on it. It's built upon two other PRs (#1558 and #1551) as well as issue #1526. 

### What should a reviewer concentrate their feedback on?

- [x] Move out the Travis part?
- [x] Add a new part about GitHub actions?
- [x] Re-arrange the CI chapter and delete unneeded sections?
- [ ] Everything looks ok?

Update 

TODO: 
- [x] The images need captions, references 
- [x] Some images I created on my own (How to change, reference)
- [ ] For the gif, we could set a media query ``@media (prefers-reduced-motion: no-preference)`` to display the GIF instead of the image. We need to take precaution for people with motion sickness/sensitivity.
- [x] Maybe add use cases and code coverage, remove the part about secret and containers
- [x] Need references at the end 
- [ ] Needs to change/revise the limitations and best practices + the checklist
- [x] Link into the documentation within the text, and for more options.
- [ ] Show GIF for a breaking action 
- [ ] Maybe add a use case for deployment

### Acknowledging contributors

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [x] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file:

### CODE REVIEW

**Comment by BatoolMM** [comment]
TODO: 
- [ ] The images need captions, references 
- [ ] Some images I created on my own (How to change, reference)
- [ ] For the gif, we could set a media query ``@media (prefers-reduced-motion: no-preference)`` to display the GIF instead of the image. We need to take precaution for people with motion sickness/sensitivity.
- [ ] Maybe add use cases and code coverage, remove the part about secret and containers
- [ ] Need references at the end 
- [ ] Needs to change/revise the limitations and best practices + the checklist

**Comment by bsipocz** [comment]
OK, rebase is done. The only difference content-wise I see with your previous branch is that the travis file is now removed (you had it removed, then it came back in a conflicting side branch, overall based on the discussion and commit messages I think we wanted it to be removed).

**Comment by BatoolMM** [comment]
Thank you @bsipocz -  yes, the travis files are removed.

**Comment by BatoolMM** [comment]
This is TODO list:
- [ ] The images need captions, references 
- [ ] Some images I created on my own (How to change, reference)
- [ ] For the gif, we could set a media query ``@media (prefers-reduced-motion: no-preference)`` to display the GIF instead of the image. We need to take precaution for people with motion sickness/sensitivity.
- [ ] Maybe add use cases and code coverage, remove the part about secret and containers
- [ ] Need references at the end 
- [ ] Needs to change/revise the limitations and best practices + the checklist
- [ ] Link into the documentation within the text, and for more options.
- [ ] Show GIF for a breaking action 
- [ ] Maybe add a use case for deployment

**Comment by bsipocz** [comment]
@BatoolMM - you've pulled in duplicated commits :)

File: book/website/reproducible-research/ci/ci-options.md [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
+
+Integrating code frequently does not, by itself, offer any guarantees about the quality of the new code or functionality. This leads us to the second aspect of CI. When a developer merges code into the main repository, automated processes build a working version of the project. Afterwards, test suites are run against the new build to check whether any bugs were introduced. If either the build or the test phase fails, the team is alerted so that they can work to fix the problem. It is easier to fix a bug in something you wrote a few minutes ago than something you wrote yesterday (or last week, or last month).
+
+By ensuring that your code is built and tested regularly CI helps researchers to demonstrate that their code does what it claims to do, and that it does so correctly. Typically, continuous integration servers will also allow build-and-test jobs to run at specific times, so a [CRON](https://en.wikipedia.org/wiki/Cron)-like, nightly-build-and-test, can be done, as well as a build-and-test job run on-demand.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-options.md:19]:
```suggestion
By ensuring that your code is built and tested regularly CI helps researchers to demonstrate that their code does what it claims to do, and that it does so correctly. Typically, continuous integration servers will also allow build-and-test jobs to run at specific times, so a [cron job](https://en.wikipedia.org/wiki/Cron), nightly-build-and-test, can be done, as well as a build-and-test job run on-demand.
```

File: book/website/reproducible-research/ci/ci-options.md [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
+
+
+### What are the options for CI service providers?
 
 There are many CI service providers, such as GitHub Actions and Travis CI. Each of these services has its own advantages and disadvantages. In this section we provide a brief overview with links to examples to help you select the most suitable one for you.
 
- - [Travis CI](https://travis-ci.com/), for some examples see [the next section](#What_is_Travis_and_how_does_it_work) and the [Travis tutorial](https://docs.travis-ci.com/user/tutorial/).
  - [GitHub Actions](https://help.github.com/en/actions), for some examples see the [language and framework guides](https://help.github.com/en/actions/language-and-framework-guides) and [this tutorial](https://github.com/NLESC-JCER/ci_for_science#-github-actions).
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-options.md:26]: maybe we can link to our docs here too.

File: book/website/reproducible-research/ci/ci-github-actions.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
+(rr-ci-ghactions)=
+# Continuous Integration with GitHub Actions
+
+This section will walk you through the basic setup of continuous integration (CI) using **GitHub Actions (GHA)**. GHA is a task automation system fully integrated with GitHub. In other words, it is an API that orchestrates any workflow based on any event. Although there are many CI service providers, GHA makes it easier than ever to incorporate CI into your repositories. It provides a flexible way to automate nearly every aspect of your project workflow. Here are just a few examples of use cases of GitHub Actions:
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-github-actions.md:4]: maybe link to othe main github actions documentation again from the first mention of github actions

Comment (bsipocz) [reply book/website/reproducible-research/ci/ci-github-actions.md:4]: Does API defined elsewhere? if not maybe rephrase that sentence

File: book/website/reproducible-research/ci/ci-github-actions.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+- Automated testing of the software
+- Generate reports of the status of any changes in the repository
+- Responding to workflow triggers using labels, issues, `@` mentions, and more
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-github-actions.md:8]:
```suggestion
- Responding to workflow triggers using labels, issues, special mentions, and more
```

File: book/website/reproducible-research/ci/ci-github-actions.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+- Triggering code reviews and pull requests
+- Managing branches
+
+GitHub Actions are event-driven, which means it responds to any event (`PR` created, issue created, ...) and triggering an action (label, test, sort, etc...). Any collection of these actions is called a workflow. A more detailed description of this GitHub-related Vocabulary is described in the next section.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-github-actions.md:12]:
```suggestion
GitHub Actions are event-driven, which means it responds to any event (pull request (PR) created, issue created, ...) and triggering an action (label, test, sort, etc...). Any collection of these actions is called a workflow. A more detailed description of this GitHub-related Vocabulary is described in the next section.
```

**Review by bsipocz — COMMENTED** [review]
half made review

File: book/website/reproducible-research/ci/ci-github-actions.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+A **step** represents one effect of the automation. A step could be defined as a GitHub Action, or another unit, like printing something to the console.
+
+#### 4. Actions
+
+A GitHub **Action** is a piece of automation written in a way that is compatible with workflows. Actions can be written by [GitHub](https://github.com/actions), by the open source [community](https://github.com/sdras/awesome-actions), or you can write them yourself!
+
+## Getting started with GitHub Action
+
+GitHub Actions uses YAML syntax and stored in a directory called `.github/workflows` in the repository. You can either use a templated workflow or create your own.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-github-actions.md:50]:
```suggestion
GitHub Actions uses YAML syntax for their workflow configuration files which are stored in a directory called `.github/workflows` in the repository. You can either use a templated workflow or create your own.
```

File: book/website/reproducible-research/ci/ci-github-actions.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+
+#### 1- Using GitHub Actions template
+
+If you want to get started with GitHub Actions, you can start by clicking the "Actions" tab in the repository where you want to create a workflow, as shown below. Under the "Actions" tab, you will find popular CI workflows, which can help deploy or automate some tasks in the repository.
+
+```{figure} ../../figures/gifs/start_ghactions.gif
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-github-actions.md:57]: I love this gif so much, super useful!

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Building a Block of a Github Actions
+
+As described previously, workflow files use YAML syntax, which has either a `.yml` or `.yaml` file extension. If you're new to YAML and want to learn more, {ref}`see our section about YMAL<rr-renv-yaml>`. This workflow must be stored in the `.github/workflows` directory of your repository.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:3]:
```suggestion
As described previously, workflow files use YAML syntax, which has either a `.yml` or `.yaml` file extension. If you're new to YAML and want to learn more, {ref}`see our section about YMAL<rr-renv-yaml>`.  Workflows must be stored in the `.github/workflows` directory of your repository.
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+The `on` field tells GHA when to run. For example, we can run the workflow anytime there's a `push` or a `pull`.
+```
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+```
+There are many events which can be used to trigger a workflow. You can explore them [here](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions).
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:38]: overall it's good practice to describe what you would find in a link rather than using generic words, e.g. `here`, `click` etc. Now, the book has quite a few `here`, so it's not a big deal in this PR either. (I'm very much nit picking, I'm sorry).

```suggestion
There are many events which can be used to trigger a workflow. You can explore them [in the official GHA reference documentation](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions).
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+**3. jobs and steps**
+
+This block defines the core component of an Action workflow. Workflows are made of `jobs`. Every job also needs a specific host machine on which to run, the `runs-on:` field is how we specify it. The template workflow is running the `build` job in the latest version of Ubuntu, a Linux-based operating system.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:42]:
```suggestion
This block defines the core component of an Actions workflow. Workflows are made of `jobs`. Every job also needs a specific host machine on which to run, the `runs-on` field is how we specify it. Our example workflow is running the `build` job in the latest version of Ubuntu, a Linux-based operating system.
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+```
+jobs:
+  build:
+  runs-on: ubuntu-latest
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:47]:
```suggestion
jobs:
  build:
    runs-on: ubuntu-latest
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  steps:
+    - uses: actions/checkout@v2
+  test:
+    - name: npm install
+      run: |
+        npm install
+        npm test
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:67]: I think you need the `steps` into `test:`, too.

```suggestion
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
  test:
    steps:
    - name: npm install
      run: |
        npm install
        npm test
```
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+```
+
+The most basic action is `uses: actions/checkout@v2`. This uses a community action called [`checkout`](https://github.com/actions/checkout) to allow the workflow to access the contents of the repository. All the steps of a job run sequentially on the runner associated with the job. By default, if a step fails, the subsequent steps of the job are skipped. Each run keyword represents a new process and shell in the runner environment. When you provide multi-line commands, each line runs in the same shell
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:70]:
```suggestion
The most basic action is `uses: actions/checkout@v2`. This uses a GitHub provided action called [`checkout`](https://github.com/actions/checkout) to allow the workflow to access the contents of the repository. All the steps of a job run sequentially on the runner associated with the job. By default, if a step fails, the subsequent steps of the job are skipped. Each run keyword represents a new process and shell in the runner environment. When you provide multi-line commands, each line runs in the same shell.
```

File: book/website/reproducible-research/ci/ci-building-gh-actions.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+Providing a comprehensive guide of all the available options is beyond the scope of this overview, and instead we would urge you to study the CI configuration of well established open source projects.
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-building-gh-actions.md:72]:
```suggestion
Providing a comprehensive guide of all the available options is beyond the scope of this overview, and instead, we would urge you to study [official reference documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions) and/or the CI configuration open-source projects references in the previous section.
```

File: book/website/reproducible-research/ci/secrets.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
+## Pro Tips
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/secrets.md:1]: I'm not sure what tips should we give here. So what do you think removing this file altogether, and rather open an issue that the current chapter could be extended with more info about secrets or using containers, or just as well writing about CD

File: book/website/reproducible-research/ci/ci-coverage.md [hunk 16]
```diff
@@ -0,0 +1,2 @@
+## Adding Code Coverage to your CI Workflow
```
Comment (bsipocz) [inline book/website/reproducible-research/ci/ci-coverage.md:1]: I think this section should be moved to the "Code testing" chapter as a specific type of test. For this CI chapter then we could mention the different actions/integrations that use a coverage service to render the result of the coverage runs. (e.g. codecov, coveralls, but there are more listed in GH's marketplace).

**Review by bsipocz — COMMENTED** [review]
This review is now full of nitpicking, but nothing major. I think we are very close to merge it, it already provides a significant improvement compared to what currently is available, so we shouldn't hold it up as PR any longer.

The future ideas then can be opened as an issue for follow-up PRs.

**Comment by bsipocz** [comment]
The latin CI fix has been merged, so this should get the all green after a rebase picks up those changes.

**Comment by bsipocz** [comment]
This was ready to get merged content-wise, I think. So how about we rebase and get all the checks happy and then merge? There is really no need to keep it as WIP any longer, but get it in the book and if anything needs more tuning, open a follow-up PR for it.

**Comment by bsipocz** [comment]
(I'm happy to rebase and get the checks sorted if you prefer)

**Review by bsipocz — APPROVED** [review]
This was ready to get merged content-wise, I think. So how about we rebase and get all the checks happy and then merge? There is really no need to keep it as WIP any longer, but get it in the book and if anything needs more tuning, open a follow-up PR for it.

**Comment by BatoolMM** [comment]
Thank you @bsipocz - Please feel free to rebase it. Always grateful to you!

**Comment by bsipocz** [comment]
All green. I take the thumbs up on my approval comment as a go ahead with the merge. 🎉

**Comment by bsipocz** [comment]
Thank you @BatoolMM for leading this much-needed update!
